### PR TITLE
Implement custom, opt-in hit-testing handlers for objects

### LINKED ISSUE #1272: Pointer events don't go through masked parts of objects
I was playing with `dev-7.0` today and I noticed that the color picker now seems to work with a transparent center, which was great to see!

The problem is that I cannot click through the center of the color picker. I'm sure that this is because input events are ignoring whatever is masking out the color picker's center.

I'd like to add this functionality, but I'm struggling to figure out where the color picker masking is being done. Does anyone know?

You can see this behavior on a 480x272 screen with the following code:

```c
   lv_test_theme_1(lv_theme_get_current());
   lv_coord_t pickerSize = 200;

    static lv_style_t styleMain;
    lv_style_copy(&styleMain, &lv_style_plain);
    styleMain.line.width = 30;
    styleMain.body.radius = 5;
	styleMain.body.main_color = styleMain.body.grad_color = LV_COLOR_WHITE;

    static lv_style_t styleIndicator;
    lv_style_copy(&styleIndicator, &lv_style_pretty);
    styleIndicator.body.border.color = LV_COLOR_WHITE;
    styleIndicator.body.opa = LV_OPA_COVER;
    styleIndicator.body.border.opa = LV_OPA_COVER;

    lv_obj_t * scr = lv_scr_act();

    lv_obj_t * colorPicker = lv_cpicker_create(scr, NULL);

    lv_obj_set_size(colorPicker, pickerSize, pickerSize);
    lv_cpicker_set_type(colorPicker, LV_CPICKER_TYPE_DISC);
    lv_obj_align(colorPicker, NULL, LV_ALIGN_CENTER, 0, 0);
    lv_cpicker_set_style(colorPicker, LV_CPICKER_STYLE_MAIN, &styleMain);
    lv_cpicker_set_style(colorPicker, LV_CPICKER_STYLE_INDICATOR, &styleIndicator);
    lv_cpicker_set_indic_colored(colorPicker, true);
```

### PR DESCRIPTION
**Pull request by embeddedt**
Fixes #1272 in a more generic manner. Advanced hit-testing must be explicitly enabled by the user for the appropriate objects, so there is 0% performance overhead by default.

Test code:

```c
    const lv_coord_t pickerSize = 200;

	/* Set the style of the color ring */
    static lv_style_t styleMain;
    lv_style_copy(&styleMain, &lv_style_plain);
    styleMain.line.width = 30;
	/* Make the background white */
    styleMain.body.main_color = styleMain.body.grad_color = LV_COLOR_WHITE;

	/* Set the style of the knob */
    static lv_style_t styleIndicator;
    lv_style_copy(&styleIndicator, &lv_style_pretty);
    styleIndicator.body.border.color = LV_COLOR_WHITE;
	/* Ensure that the knob is fully opaque */
    styleIndicator.body.opa = LV_OPA_COVER;
    styleIndicator.body.border.opa = LV_OPA_COVER;

    lv_obj_t * scr = lv_scr_act();

    static lv_style_t btn_pr, btn_rel;
    lv_style_copy(&btn_pr, &lv_style_btn_pr);
    lv_style_copy(&btn_rel, &lv_style_btn_rel);
    btn_pr.body.radius = LV_RADIUS_CIRCLE;
    btn_rel.body.radius = LV_RADIUS_CIRCLE;

    lv_obj_t * btn = lv_btn_create(scr, NULL);
    lv_obj_set_size(btn, 100, 100);
    lv_btn_set_style(btn, LV_BTN_STYLE_PR, &btn_pr);
    lv_btn_set_style(btn, LV_BTN_STYLE_REL, &btn_rel);
    lv_obj_align(btn, NULL, LV_ALIGN_CENTER, 0, 0);

    lv_obj_t * label = lv_label_create(btn, NULL);
    lv_label_set_text(label, LV_SYMBOL_POWER);

    lv_obj_t * colorPicker = lv_cpicker_create(scr, NULL);
    lv_obj_set_adv_hittest(colorPicker, true);
    lv_obj_set_size(colorPicker, pickerSize, pickerSize);
	/* Choose the 'DISC' type */
    lv_cpicker_set_type(colorPicker, LV_CPICKER_TYPE_DISC);
    lv_cpicker_set_preview(colorPicker, false);
    lv_obj_align(colorPicker, NULL, LV_ALIGN_CENTER, 0, 0);
	/* Set the styles */
    lv_cpicker_set_style(colorPicker, LV_CPICKER_STYLE_MAIN, &styleMain);
    lv_cpicker_set_style(colorPicker, LV_CPICKER_STYLE_INDICATOR, &styleIndicator);
	/* Change the knob's color to that of the selected color */
    lv_cpicker_set_indic_colored(colorPicker, true);
```

### CODE REVIEW

**Comment by kisvegabor** [comment]
I've 2 questions:
1. Can we use a signal instead of a new callback? The signal function could use `lv_indev_get_point(lv_indev_act, &p)` and test it. The result can be passed  to the `void * param`
2. As lvgl doesn't support ellipse drawing/handling `lv_point_within_ellipse` seems pointless to me. A function like `lv_point_is_in_rect(&area, radius, &point)` would make more sense to me.

What do you think?

**Comment by embeddedt** [comment]
I agree with `#1`.

I'm not sure about `#2`. Wouldn't it make the function significantly slower if it had to deal with rounded rectangles? Could we change it to `lv_point_is_in_circle`?

File: src/lv_core/lv_obj.h [hunk 1]
```diff
@@ -147,6 +148,7 @@ enum {
 typedef uint8_t lv_signal_t;
 
 typedef lv_res_t (*lv_signal_cb_t)(struct _lv_obj_t * obj, lv_signal_t sign, void * param);
+typedef bool (*lv_hittest_cb_t)(struct _lv_obj_t * obj, const lv_point_t * point);
```
Comment (kisvegabor) [inline src/lv_core/lv_obj.h]: Is this still required?

File: src/lv_core/lv_obj.h [hunk 2]
```diff
@@ -263,6 +266,12 @@ typedef struct
                                                ... [x]: "lv_obj" */
 } lv_obj_type_t;
 
+typedef struct _lv_hit_test_info_t
```
Comment (kisvegabor) [inline src/lv_core/lv_obj.h:268]: Can it go to `lv_obj.c`?

**Comment by kisvegabor** [comment]
It's a great idea to pass `lv_hit_test_info_t` which contains 2 separate fields for input and output data.

> I'm not sure about #2. Wouldn't it make the function significantly slower if it had to deal with rounded rectangles? Could we change it to lv_point_is_in_circle?

I think it shouldn't be that slow. It seems only following checks need to be performed:
1. Is the point in the area?
2. If the point in the corner area? (for left top corner in the `{x1,y1}, {x1+r, y1+r}` area)
3. If so `x^2 + y^2 <= r^2` can quickly tell whether the point is in or on the circle or not.

Or have I missed something?

File: src/lv_core/lv_obj.h (hunk 2 above)
Comment (embeddedt) [reply src/lv_core/lv_obj.h:268]: Objects need to be able to access the structure, so it has to be visible to them in a header file.

File: src/lv_core/lv_obj.h (hunk 1 above)
Comment (embeddedt) [reply src/lv_core/lv_obj.h]: Nope. I forgot to remove it. :slightly_smiling_face:

**Comment by embeddedt** [comment]
I agree. It's not actually too complex. I've pushed a new version which does it the way you suggested.

File: src/lv_core/lv_obj.h (hunk 2 above)
Comment (kisvegabor) [reply src/lv_core/lv_obj.h:268]: Oh, true!

File: src/lv_misc/lv_area.c [hunk 3]
```diff
@@ -208,3 +257,24 @@ void lv_area_increment(lv_area_t * a_p, const lv_coord_t amount)
 /**********************
  *   STATIC FUNCTIONS
  **********************/
+
+static bool lv_point_within_circle(const lv_area_t * area, const lv_point_t * p)
+{
+    lv_coord_t r = (area->x2 - area->x1) / 2;
+
+    /* Ellipse center */
```
Comment (kisvegabor) [inline src/lv_misc/lv_area.c]: /* Ellipse center */
 
```suggestion
    /* Circle center */
```

**Review by kisvegabor — COMMENTED** [review]
Looks good!

Only a minor typo:

File: src/lv_misc/lv_area.c (hunk 3 above)
Comment (embeddedt) [reply src/lv_misc/lv_area.c]: I actually had it in mind to change that before committing, and then I completely forgot! Thanks!

**Comment by embeddedt** [comment]
Thanks for the review! I'll merge this now.

**Comment by kisvegabor** [comment]
Thank you for this feature :slightly_smiling_face:

**Pull request merged**